Embedded bias initial implementation

danielandresarcones opened the pull request:
This is the initial implementation of embedded bias approaches to be integrated in probeye. It provides a working implementation for the computation of inverse problems with embedded bias.
- Currently the bias embedding must be implemented manually in the forward model. Ideally, this should not be the case and the bias should be specified in the parameters and variables description.
- The additions take care of the MCMC solver with stochastic forward model response and introduce four embedded likelihood formulations: ABC Moment matching, Global Moment Matching, Relative Global Moment Matching and Independent Normal. The details on the implementations will eventually be published.
- Kennedy and O'Hagan-type formulations are not implemented in this PR and have been eliminated from the branched code.
- Tests and one simple example for the documentation have been added. 
- `chaospy` has been added as a dependency. Probably should be an optional one, but kept as compulsory for now.
- `mypy` was updated to the latest version

danielandresarcones committed:
Modified Likelihood models for KOH

danielandresarcones committed:
Start modified KOH solver

danielandresarcones committed:
Modified KOH solver

danielandresarcones committed:
New biased inverse

danielandresarcones committed:
Running KOH solver for extension

danielandresarcones committed:
biased inverse problems and solver for KOH

danielandresarcones committed:
Updated OGP solver

danielandresarcones committed:
Scaling of KOH for extended problems

danielandresarcones committed:
Correct bug solver koh

danielandresarcones committed:
Added residuals scaling

danielandresarcones committed:
Modified solver

danielandresarcones committed:
Updated ogp solver

danielandresarcones committed:
Merge pull request #2 from BAMresearch/main

Updated to last state

danielandresarcones committed:
Formating

danielandresarcones committed:
Merge remote-tracking branch 'fork/koh_implementation' into koh_implementation

danielandresarcones committed:
Update black

danielandresarcones committed:
Added ABC Likelihood with MCI and PCE

danielandresarcones committed:
Reference ABC Likelihood

danielandresarcones committed:
Updated PCE solver for surrogate selection

danielandresarcones committed:
Fix Likelihood models

danielandresarcones committed:
Modified scipy solver to always pass experiment name

danielandresarcones committed:
Merge branch 'koh_implementation' of https://github.com/danielandresarcones/probeye into koh_implementation

danielandresarcones committed:
Noise in moment-matching likelihood

danielandresarcones committed:
Updated embedded likelihoods

danielandresarcones committed:
Fix Gamma high n

danielandresarcones committed:
Added guard against NaN

danielandresarcones committed:
Fixed bug var to std from pce

danielandresarcones committed:
Added tests for embedded bias

danielandresarcones committed:
Add bias example to docs

danielandresarcones committed:
Merge branch 'main' into embedded_bias

danielandresarcones committed:
Black and others

danielandresarcones committed:
Added version and dependencies

danielandresarcones committed:
Changes from black

danielandresarcones committed:
Changes for mypy v1.11.2

danielandresarcones committed:
Changed base likelihood descritptor for embedded lms

danielandresarcones committed:
CI changes for mypy v1.11.2

danielandresarcones committed:
Deleted bias

danielandresarcones committed:
Minor fixes

danielandresarcones committed:
Updated tests and fix Avoid copy.deepcopy (restriction in FEniCS) #90

danielandresarcones committed:
Black format

danielandresarcones commented:
This update also implements the proposed solution of Issue #90 .

danielandresarcones commented:
Tests cover is not 100% on the new implementations as there probably will be changes in the definition of the bias in the near future. The lack of cover is precisely on that front, the checks that some parameter definitions are correctly imported to the models. Nevertheless, they can be tested if considered necessary.

danielandresarcones committed:
Fixed bias tests

danielandresarcones committed:
Faster embedded bias examples

joergfunger commented on docs/examples/bias.py:
```diff
@@ -0,0 +1,264 @@
+# model response is calculated using PCEs with pseudo-spectral decomposition.
+
+
+class LinearModel(ForwardModelBase):
```
This is an elegant way of using the flexibility with the Forward model output to return directly a distribution rather then sensor output. But that means the ForwardModel is not compatible with other deterministic ForwardModels. Would it make sense to Create a wrapper for the StochasticForwardModels, that takes as input a deterministic forward model (as well as the definition of the embedding and the parameters of the PCE) and then wraps that pce expansion as a blackbox, anyone else could then easily add his/her model, and then you could also check, whether the Forward model send to the embedded solver is actually a stochastic forward model.

danielandresarcones replied:
Agree, it will be the case for the reimplementation of probeye. For now it remains as something the user must consider when building their forward models.

danielandresarcones committed:
Restart run PCE provides inference data

danielandresarcones committed:
black

danielandresarcones committed:
Added sampled biases

danielandresarcones committed:
Undoing expansion definition. Modifying deepcopy

danielandresarcones committed:
Modified ABC for tolerance

aradermacher commented on docs/examples/bias.py:
```diff
@@ -0,0 +1,264 @@
+        b_dist = chaospy.Normal(0.0, b)
+
+        # generate the polynomial chaos expansion
+        expansion = chaospy.generate_expansion(self.pce_order, bias_dist)
```
It should be b_dist instead of bias_dist

danielandresarcones replied:
I missed that totally, thanks!

aradermacher reviewed (commented):
Thanks Daniel, 
I had problems running the example file, because of a typo (see comment)
Furthermore, during creating a new env I run into an error with python 3.13. Changing to 3.10 solved it.

danielandresarcones committed:
Update typo in example

bias_std -> b_dist

danielandresarcones commented:
Do you remember what whas the error with python 3.13? I think probeye had some problems with it at some point, not necessarily the embedded bias implementation only.

aradermacher commented:
> Do you remember what whas the error with python 3.13? I think probeye had some problems with it at some point, not necessarily the embedded bias implementation only.

Yes it was something with the required packages (but I didn't saved the message sorry). So I think it's a general problem of probeye and not with your implementation.

Commit:
try pool

danielandresarcones committed:
Modified noise in GMM and RGMM

danielandresarcones committed:
Merge branch 'embedded_bias' of https://github.com/BAMresearch/probeye into embedded_bias

Commit:
finalized pool

Commit:
implement parallel sample computation using Pool

Commit:
black formatting

Commit:
black formatting

Commit:
parallel sampling with Pool

Commit:
fix mypy check

Commit:
fix black style

Commit:
delete parallel stuff - moved to own branch

danielandresarcones committed:
Introduced RelativeScaledGMMError

danielandresarcones committed:
Merge branch 'embedded_bias' of https://github.com/BAMresearch/probeye into embedded_bias

danielandresarcones committed:
Merge remote-tracking branch 'origin/embedded_parallel' into embedded_bias

danielandresarcones committed:
parallelization with pool optional in emcee

danielandresarcones committed:
Add optional parallel sampling option to EmbeddedPCESolver

danielandresarcones committed:
Modified bias likelihood models

danielandresarcones committed:
Simplified likelihood and error definition with bias

danielandresarcones committed:
Adapt tests to new formulation

danielandresarcones committed:
Solved deepcopy Avoid copy.deepcopy (restriction in FEniCS) #90

danielandresarcones committed:
Merge pull request #117 from BAMresearch/embedded_bias_paper

Embedded bias paper

danielandresarcones commented:
This version is the first step towards #115. Main changes:

- Parallelization of walkers in emcee-based solvers.
- Simplified creation of new likelihoods (bias likelihoods were created in the new way). They can be defined without having to deal with the correlation features
- Fixed the bias implementations
- Solved copy/deepcopy issue from #90

danielandresarcones committed:
Fixed parallel and RGMM

danielandresarcones committed:
Merge pull request #118 from BAMresearch/embedded_bias_paper

Fixed parallel and RGMM

danielandresarcones committed:
Fixed global logprob

danielandresarcones committed:
Merge pull request #119 from BAMresearch/embedded_bias_paper

Fixed global logprob

joergfunger commented on docs/examples/bias.py:
```diff
@@ -0,0 +1,261 @@
+
+    def response(self, inp: dict) -> dict:
+        x = inp["x"]
+        m = inp["a"]
```
the naming is not clear, (I guess m stands for mean, b for bias), but some short documentation would be nice

danielandresarcones replied:
Modified all the examples for consistent naming. Now "a" is the slope and "b" is the cross with the y-axis.

joergfunger commented on docs/examples/plot_correlation_1D.py:
```diff
@@ -198,14 +198,16 @@ def response(self, inp: dict) -> dict:
 # this is for using the scipy-solver (maximum likelihood estimation)
-scipy_solver = MaxLikelihoodSolver(problem, show_progress=False)
-max_like_data = scipy_solver.run()
+# scipy_solver = MaxLikelihoodSolver(problem, show_progress=False)
```
why is that uncommented?

danielandresarcones replied:
Forgotten from debugging, removed

joergfunger commented on docs/examples/plot_correlation_1D.py:
```diff
@@ -198,14 +198,16 @@ def response(self, inp: dict) -> dict:
 # run-method, which returns its result data in the format of an arviz inference-data
 # object (except for the scipy-solver). Let's now take a look at the emcee-solver.
 
+problem.forward_models.pop("LinearModel")
```
Why is the model first removed and then added?

danielandresarcones replied:
Forgotten from debugging, removed

joergfunger commented on probeye/definition/inverse_problem.py:
```diff
@@ -92,11 +92,14 @@ def __init__(
         # problem can have multiple forward models; the keys in this dict are the
         # forward model names, while the values are the forward model objects (check out
         # the script forward_model.py)
+        # the internal forward models list enables reusing the same problem in different
```
not sure I understand that, why do we need internal forward models? Why don't you define difference inference problems when using different solvers?

danielandresarcones replied:
See below. It is so you can use a shallow copy to re-use the forward model and not needing to do the pop+readding trick or defining a totally new problem. Just for consecutive uses of the same forward model (relevant for reinitiating a run)

joergfunger commented on probeye/inference/bias/likelihood_models.py:
```diff
@@ -0,0 +1,431 @@
+            self.weight_std = 1.0
+
+        # Load the standard deviations and noise values
+        std_model, std_meas, stds_are_scalar = self.std_values(prms)
```
not sure I understand what that methods does? Could you add some docs?

danielandresarcones replied:
Do you mean extending the docstring of the method or writing a comment there in the application? It is a method from the base class, it is defined in ScipyLikelihoods

joergfunger commented on probeye/inference/bias/likelihood_models.py:
```diff
@@ -0,0 +1,431 @@
+
+        # Calculate the intermediate statistics
+        mean_residual = np.mean(residual_vector)
+        variance_residual = np.var(residual_vector)
```
I think this is assuming the mean to be known, but since it is estimated, you have to divide by n-1 rather than n, use ddof=1 in this case

danielandresarcones replied:
Updated in the new version

joergfunger commented on probeye/inference/bias/likelihood_models.py:
```diff
@@ -0,0 +1,431 @@
+        variance_residual = np.var(residual_vector)
+        mean_var = np.mean(np.square(response_vector[1]))
+        sample_variance = mean_var + variance_residual + variance
+        population_variance = (
```
I think the variable notation is not matching with the paper, it would be nice if that could be improved (e.g. population variance is defined differently in the paper in eq. 31)

danielandresarcones replied:
Updated in the new version

joergfunger commented on probeye/inference/bias/likelihood_models.py:
```diff
@@ -0,0 +1,431 @@
+        n = len(residual_vector)
+
+        # Calculate the intermediate statistics
+        sigma_model_population = np.sqrt(
```
not sure I see this in eq. 42

danielandresarcones replied:
Updated in the new version

joergfunger commented on probeye/inference/bias/likelihood_models.py:
```diff
@@ -0,0 +1,431 @@
+        )
+        sigma_model_sample = np.sqrt(np.square(response_vector[1]) + variance)
+        population_variance = np.var(np.divide(data, sigma_model_population)) + 1
+        sample_variance = np.var(np.divide(response_vector[0], sigma_model_sample)) + 1
```
not sure I would I see the similarity to eq. 47 in the paper

danielandresarcones replied:
Updated in the new version

joergfunger commented on probeye/inference/bias/likelihood_models.py:
```diff
@@ -0,0 +1,431 @@
+
+class RelativeScaledGlobalMomentMatchingModelError(EmbeddedLikelihoodBaseModel):
+    """
+    This class implements the relative global moment matching likelihood model.
```
not sure what the difference between the scaled global moment matching and the unscaled version is? Is there a reference in the paper for that likelihood?

danielandresarcones replied:
I removed all the likelihoods that are not being used in the paper, they were just variations that I tried at some point.

joergfunger commented on probeye/inference/bias/likelihood_models.py:
```diff
@@ -0,0 +1,431 @@
+        n = len(residual_vector)
+
+        # Calculate the intermediate statistics
+        sigma_model_population = np.sqrt(
```
is that eq. 42?

danielandresarcones replied:
See updated version

joergfunger commented on probeye/inference/bias/likelihood_models.py:
```diff
@@ -0,0 +1,431 @@
+        sigma_model_sample = np.sqrt(np.square(response_vector[1]) + variance)
+        population_variance = (
+            np.var(np.divide(residual_vector, sigma_model_population)) + np.mean(1 - np.divide(variance, np.square(sigma_model_population)+variance))
+        )
```
not sure I can follow here with the equations from the paper

danielandresarcones replied:
Updated

joergfunger commented on probeye/inference/bias/likelihood_models.py:
```diff
@@ -0,0 +1,431 @@
+        if stds_are_scalar:
+            ll -= 0.5 * np.sum(
+                np.square(np.divide(residual_vector, sigma_model_sample))
+                + np.log(2 * np.pi * np.square(sigma_model_sample))
```
where is the -ny/2 from eq. 28? and where is the inverse in the last term?

danielandresarcones replied:
See updated version

joergfunger commented on probeye/inference/scipy/solver.py:
```diff
@@ -105,7 +105,7 @@ def _translate_forward_models(self):
 
             # finally, add the forward model to the problem
-            self.problem.forward_models[fwd_name] = forward_model
+            self.problem.internal_forward_models[fwd_name] = forward_model
```
see above not sure what these internal forward models are

danielandresarcones replied:
These are here to solve the copy/deepcopy problem. See #90 for full detail, but the problem is triggered when several solvers access the problem sequentially. FEniCS(X) models cannot be deepcopied, so we need to change them all to copies. Then, the forward model definition is overwritten in the solver initialization, which means that the next solvers will  see only the hull and throw an error. By assigning the translated forward model to the list "internal_forward_models" instead of overwriting "forward_models" and adapting the solvers accordingly, we still preserve the forward model definition after the first call, and is the "internal" one that is overwritten. The deepcopies are therefore not needed as the forward_model definition is still available after running the problem and a new "internal" forward model is created and evaluated from it.

joergfunger replied:
I see, thanks. Just one add-on question. Is it required to store the these internal forward models on the level of the problem definition, or might it be sufficient to store them only locally in the solver, where these copies are actually required? In addition, the deep copy was done to prevent the copies to interact if they run in parallel, which as far as I understand here is not intended (the copy would result in identical objects e.g. for the function spaces that are shared between the different copied models). Wouldn't that mean we could just work on the original model, i.e. preventing parallel execution and in this case avoid any copy?

danielandresarcones replied:
We could store them in the solver, but that means modifying the way the likelihood access to them as well (right now it reads from the problem). We could avoid the copy by creating new objects as long as we do not create a hull anymore, which means that definition and execution of the forward model would not be decoupled anymore. If that is still necessary is something we would have to discuss.

joergfunger replied:
I guess this is something for the redesign of the interface, lt's tackle that in the subsequent redesign.

joergfunger commented on probeye/inference/solver.py:
```diff
@@ -31,7 +31,7 @@ def __init__(
 
         # attributes from arguments
-        self.problem = cp.deepcopy(problem)
+        self.problem = cp.copy(problem)
```
why is a copy here sufficient?

danielandresarcones replied:
See above.

joergfunger reviewed (changes requested):
nice work, thanks. There are a couple of remarks or thinks that I have difficulties to relate to the current paper. It would be nice if the naming conventions in the paper and the source code are somehow overlapping, such that the paper serves as a scientific documentation to the source code.

danielandresarcones committed:
Eliminated likelihood models not present in paper

danielandresarcones committed:
Modified MomentMathching to not silent fail heteroscedastic noise

danielandresarcones committed:
Updated GMM likelihood with paper formulation

danielandresarcones committed:
Reorganize likelihoods like in paper

danielandresarcones committed:
Refactor likelihood model handling and internal forward models in solvers and tests

danielandresarcones committed:
Merge branch 'embedded_bias' of https://github.com/BAMresearch/probeye into embedded_bias

danielandresarcones committed:
Cleaned likelihood_models after merge

danielandresarcones committed:
Updated RGMM likelihood with paper formulation

danielandresarcones committed:
Updated equation numbers

danielandresarcones committed:
Fix variance calculation and update test assertions for consistency

danielandresarcones committed:
Bias example - improve variable naming consistency

danielandresarcones committed:
Fix plot_correlation_1D

danielandresarcones committed:
Comment ML solver

danielandresarcones committed:
Black formatting

joergfunger reviewed (approved):
Thanks
